src/utils/slugify.ts

```diff
@@ -0,0 +1,36 @@
+export function slugify(input: string): string {
+  // See src/utils/__test__/slugify.test.ts for examples
+
+  const result = removeDiacritics(input).replaceAll(' ', '-').toLowerCase();
+
+  const unhandledCharacters = result.match(/[^abcdefghijklmnopqrstuvwxyz0123456789_.-]/g);
+  if (unhandledCharacters) {
+    const sortedUniqueCharacters = Array.from(new Set(unhandledCharacters)).sort();
+    throw new UnhandledCharactersError(sortedUniqueCharacters);
+  }
+
+  return result;
+}
+
+function removeDiacritics(input: string): string {
+  /*
+   Normalization form decomposition (NFD) splits characters like into their
+   [combining diacritical mark](https://www.unicode.org/charts/PDF/U0300.pdf) and the character which is being modified
+   by the diacritic. This way we can remove the macron from "ā", the accent from "é", and the like.
+   */
+  const combiningDiacriticalMarks = /[\u0300-\u036F]/g;
+  return input.normalize('NFD').replaceAll(combiningDiacriticalMarks, '');
+}
+
+class UnhandledCharactersError extends Error {
```

Comment:
We dont generally subclass errors due to how the instanceof checking for them can often fail

Reply:
Generally you just hope the "name"/"message" contains enough information to give the cause.

For more complex errors its a good idea to wrap them with additional information

```
try {

} catch(e) {
 throw new Error(`Failed to parse: ${url.href}`, { cause: e })
}
```

Trying to determine error types at runtime is very hit and miss as `catch(e: unknown)` since you are able to throw anything, `throw 1` is a perfectly valid throw

```
> try {throw 1} catch(e) { console.log(typeof e) }
number
```


The MDN docs for Error have a example with structured error causes, which would be a better way of handling this.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause#providing_structured_data_as_the_error_cause

Reply:
`tsc` doesn't like that:

```
❯ npm run build && npm run test && npm run lint -- --fix

> @linzjs/argo-tasks@2.18.2 build
> tsc

src/utils/slugify.ts:15:80 - error TS2554: Expected 0-1 arguments, but got 2.

15     throw new Error(`Unhandled characters: ${formattedCharacters.join(', ')}`, {
                                                                                  ~
16       cause: {characters: sortedUniqueCharacters}
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
17     });
   ~~~~~


Found 1 error in src/utils/slugify.ts:15
```

Reply:
ahh argo-tasks is currently defaulting back to node16 typings, we need to add

```json
    "lib": ["ES2022"],
    "target": "ES2022",
```

to the `tsconfig.json#compilerOptions` which will switch it to node18 types, when we move to node20 it should change to ES2023

Reply:
What else do I need to do? Adding those doesn't change the `tsc` result.

Reply:
Just checked out your branch updated tsconfig to

```typescript
{
  "extends": "@linzjs/style/tsconfig.base.json",
  "compilerOptions": {
    "lib": ["ES2022"],
    "target": "ES2022",
    "outDir": "build"
  }
}
```

`tsc` works then switching back to 2021 `tsc` dies when I add a `throw new Error("foo", { cause: "bar" })`

Reply:
Didn't see the JSONPath in your reference; you're right.

Reply:
This should have been changed when we migrated from node16 to node18.

Reply:
Fixed.